File: dart/lib/bebop_handlers/bebop_handler.dart
import './err_msg.dart';
import './session.dart';
import './exit.dart';
import './player_state.dart';
import './ghost_move.dart';
import './player_joined.dart';
import './ready.dart';
import './init_state.dart';
import './reset.dart';
import './reward.dart';
import './game_over.dart';
import './new_level.dart';
import './new_game.dart';

abstract class BebopHandler {
  BebopHandler();

  factory BebopHandler.fromClassName(String className) {
    switch (className) {
      case 'ErrorMsg':
        return ErrorMsgHandler();
      case 'SessionMsg':
        return SessionMsgHandler();
      case 'ExitMsg':
        return ExitMsgHandler();
      case 'JoinMsg':
        return null; // not used by client
      case 'PlayerState':
        return PlayerStateMsgHandler();
      case 'GhostMoveMsg':
        return GhostMoveMsgHandler();
      case 'PlayerJoinedMsg':
        return PlayerJoinedMsgHandler();
      case 'ReadyMsg':
        return ReadyMsgHandler();
      case 'InitState':
        return InitStateMsgHandler();
      case 'ResetMsg':
        return ResetMsgHandler();
      case 'ReconnectMsg':
        return null; // not used by client
      case 'GameOverMsg':
        return GameOverMsgHandler();
      case 'NewLevelMsg':
        return NewLevelMsgHandler();
      case 'NewGameMsg':
        return NewGameMsgHandler();
      case 'RewardMsg':
        return RewardMsgHandler();
      default:
        throw UnimplementedError('Unknown bebop schema class $className');
    }
  }

  void handleMessage(dynamic sender, dynamic message) {
    throw UnsupportedError('BebopHandler must be extended!');
  }
}

File: typescript/src/bebop_handlers/bebopHandler.ts
import ErrorMsgHandler from "./err";
import ExitMsgHandler from "./exit";
import GameOverMsgHandler from "./gameOver";
import GhostMoveMsgHandler from "./ghostMove";
import InitStateMsgHandler from "./initState";
import NewGameMsgHandler from "./newGame";
import NewLevelMsgHandler from "./newLevel";
import PlayerJoinedMsgHandler from "./playerJoined";
import PlayerStateMsgHandler from "./playerState";
import ReadyMsgHandler from "./ready";
import ResetMsgHandler from "./reset";
import RewardMsgHandler from "./reward";
import SessionMsgHandler from "./session";
import StrikeMsgHandler from "./strikeMsg";

export default abstract class BebopHandler {

    static fromClassName(className: string): BebopHandler | null {
        switch (className) {
            case 'ErrorMsg':
              return new ErrorMsgHandler();
            case 'StrikeMsg':
              return new StrikeMsgHandler();
            case 'SessionMsg':
              return new SessionMsgHandler();
            case 'ExitMsg':
              return new ExitMsgHandler();
            case 'JoinMsg':
              return null; // not used by client
            case 'PlayerState':
              return new PlayerStateMsgHandler();
            case 'GhostMoveMsg':
              return new GhostMoveMsgHandler();
            case 'PlayerJoinedMsg':
              return new PlayerJoinedMsgHandler();
            case 'ReadyMsg':
              return new ReadyMsgHandler();
            case 'InitState':
              return new InitStateMsgHandler();
            case 'ResetMsg':
              return new ResetMsgHandler();
            case 'ReconnectMsg':
              return null; // not used by client
            case 'GameOverMsg':
              return new GameOverMsgHandler();
            case 'NewLevelMsg':
              return new NewLevelMsgHandler();
            case 'NewGameMsg':
              return new NewGameMsgHandler();
            case 'RewardMsg':
              return new RewardMsgHandler();
            default:
              throw new Error('Unknown bebop schema class $className');
          }
    }

    handleMessage(sender: any, message: any) {
        throw new Error("BebopHandler must be extended!");
    }
}